### PR TITLE
Customer names in reports

### LINKED ISSUE #5054: As a hub, I can give my suppliers access to my customers' name and firstname
## Description
**- As a:** enterprise manager
**- On page:** `/admin/enterprises/NAME/edit#/shop_preferences`
**- I want to be able to do:** allow my producers to see my customers names and firstnames in the following reports:

- Orders and distributors (one column: Customer name)
- Bulk Co-op (one column: Customer name)
- Order Cycle Customer Totals (one column: Customer name)
- All Customers (**name and first names are two columns on these ones**!)
- Enterprise Fee
- All Order Cycle Management  (**name and first names are two columns on this one**!)
- All Sale Tax
- Xero invoices
- All packing reports  (**name and first names are two columns on these ones**!)

**All new content must be translatable (it's not an English word but I mean no translation key missing at the end!).**

## Acceptance Criteria & Tests

*Scenario 1: Shop preferences for the hub*
1. As a hub, I can go on `/admin/enterprises/NAME/edit#/shop_preferences`
2. And see this new radio button at the bottom of the page:
image
3. By default the radio button is at NO
5. The _what's this_ tooltip displays the following tip: `Enable your suppliers to see your customers name and firstname in reports`.
4. If I switch it to YES. All Names and first names of my customer will be visible to **ALL** my suppliers. Emails and phone numbers will remain hidden (which means the word "hidden" is displayed or the column is empty).
5. I need to save at the bottom for my changes to be saved.
6. Once saved I see this green message at the top of my screen: `Enterprise "NAME" has been successfully updated!`

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #5054, #5097

- Adds a new enterprise "Shop Preference" to allow suppliers to view customer names in reports.
- Conditionally masks relevant customer names based on the new setting

#### What should we test?

Enterprises can set an "allow customer names" preference in admin enterprise edit page, under "Shop Preferences". The default selection is "disabled".

The following reports need to be tested:

- Orders and distributors report (one column: Customer name)
- Bulk Co-op reports  (one column: Customer name)
- Orders and Fulfillment - Order Cycle Customer Totals report (one column: Customer name)
- Enterprise Fees report
- Both Packing reports (name and first names are two columns on these ones!)

The test conditions are as follows:

- The current user manages a **supplier** that supplied products in some placed orders. The user should not manage the **distributor** enterprise the orders were placed with.
- The supplier enterprise has set the enterprise permission "add to order cycle" for a the distributor.
- The distributor enterprise has the "allow customer names" shop preference disabled.
- The (supplier enterprise) user looks at the reports listed above, and can not see customer names.
- The distributor enterprise then sets the "allow customer names" shop preference to enabled.
- The user managing the supplier can then view the reports listed below, and customer names are visible.

#### Release notes

Added option for distributors to allow their suppliers to see customer names in reports, for improved packing

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
I need a bit of product feedback on the messages to show to users: https://github.com/openfoodfoundation/openfoodnetwork/issues/5097#issuecomment-658786393

File: app/serializers/api/admin/enterprise_serializer.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@ class Api::Admin::EnterpriseSerializer < ActiveModel::Serializer
              :payment_method_ids, :shipping_method_ids, :producer_profile_only, :long_description,
              :preferred_shopfront_message, :preferred_shopfront_closed_message,
              :preferred_shopfront_taxon_order, :preferred_shopfront_order_cycle_order,
-             :preferred_product_selection_from_inventory_only,
+             :preferred_product_selection_from_inventory_only, :preferred_allow_customer_names,
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/serializers/api/admin/enterprise_serializer.rb:6]: These are dynamic accessors, right?. The preference is called `allow_customer_names`. When preferences are serialized they're accessed with `#preferred_<preference>`, and when accessing it from an instance you can use `#prefers_<preference>?` to check the preference is set.

File: app/models/enterprise.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@ class Enterprise < ActiveRecord::Base
   preference :shopfront_closed_message, :text, default: ""
   preference :shopfront_taxon_order, :string, default: ""
   preference :shopfront_order_cycle_order, :string, default: "orders_close_at"
+  preference :allow_customer_names, :boolean, default: false
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/models/enterprise.rb]: Updated :+1:

**Comment by Matt-Yorkley** [comment]
I wanted to a lot more here, but was told to do the minimum to cover the epic. I've updated and refactored a lot more :+1:

File: app/serializers/api/admin/enterprise_serializer.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (luisramos0) [reply app/serializers/api/admin/enterprise_serializer.rb]: I read the preferable code a little bit more and realized they are aliases for the same method:
alias_method prefers_getter_method(name), preference_getter_method(name)

so, it doesn't matter but it does look better, more readable, to use prefers for boolean like this one, I think that is the reason to have both aliases.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Thanks for this new feature! Had a look at the different scenarios in your test notes, and as well on the feature requests from @RachL .

**Before staging this PR**

- Created a profile only producer A:

image

image

- Created product B, with items in stock

- Set this producer A to act as a supplier to an enterprise, as indicated by the permissions below:
image

As a hub/enterprise manager C:
- Verified that the "allow customer names" preference does not exist on the dashboard.
- Added the product B from the supplier A to an open order cycle

As a customer D:
- visited the shop from the enterprise C (which sells the produce B from the supplier A)
- inserted the product B in the cart and checked out.

As Producer A:
-  verified that the name/surname from customer D does not appear in the reports

    Orders and distributors report (one column: Customer name - appears as HIDDEN)

image

    Bulk Co-op reports (one column: Customer name - not appearing)

image

    Orders and Fulfillment - Order Cycle Customer Totals report (one column: Customer name - not appearing)

image

    Enterprise Fees report

image

    Both Packing reports (name and first names are two columns on these ones! - appears as HIDDEN)

image


**After staging this PR**

As a hub/enterprise manager C:
- Verified that the "allow customer names" preference now exists on the dashboard, and can now be set, and that it's default preference is "disabled":

image

As Producer A:
- verified that the reports shown above look the same, and still hide the names of customers.

As a hub/enterprise manager C:
- Enabled the "allow customer names" preference 

As Producer A:
- checked all the reports again - customer names are now visible :tada: 


    Orders and distributors report (one column: Customer name is visible, email is hidden)

image

    Bulk Co-op reports (one column: Customer name - now appearing)

image

    Orders and Fulfillment - Order Cycle Customer Totals report (one column: Customer name - not appearing)

image

    Enterprise Fees report - customer name now visible

image

    Both Packing reports (name and first names are two columns on these ones! - now appearing)

image

image


Awesome work! Thank you @RachL for inception and @Matt-Yorkley for implementing :muscle: 
Ready to go.

**Pull request merged**